app/models/annotation_category.rb

```diff
@@ -30,15 +30,40 @@ def self.add_by_row(row, assignment, current_user)
     annotation_category = assignment.annotation_categories.find_or_create_by(
       annotation_category_name: name
     )
-
-    row.each do |text|
-      annotation_text = annotation_category.annotation_texts.build(
-        content: text,
-        creator_id: current_user.id,
-        last_editor_id: current_user.id
-      )
-      unless annotation_text.save
-        raise CsvInvalidLineError
+    raise CsvInvalidLineError unless annotation_category.valid?
+    # The second column is the optional flexible criterion name.
+    criterion_name = row.shift
+    if criterion_name.nil?
+      row.each do |text|
+        annotation_text = annotation_category.annotation_texts.build(
+          content: text,
+          creator_id: current_user.id,
+          last_editor_id: current_user.id
+        )
+        unless annotation_text.save
+          raise CsvInvalidLineError
+        end
+      end
+    else
+      criterion = assignment.flexible_criteria.find_by(name: criterion_name)
+      raise CsvInvalidLineError if criterion.nil?
+      annotation_category.update!(flexible_criterion_id: criterion.id)
+      row.each_slice(2) do |text_with_deduction|
+        content = text_with_deduction.first
+        deduction = text_with_deduction.second.to_f
+        if deduction > criterion.max_mark
```

Comment:
This part shouldn't be necessary. On line 62, set `deduction: deduction` rather than the min of `criterion.max_mark` and `deduction`. If the `deduction` is too large, the call to `save` will return `false`.